File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.cpp
#include <bits/stdc++.h>
using namespace std;
int dx[]{0,0,1,-1};
int dy[]{1,-1,0,0};
int solution(vector<vector<int> > maps)
{
    int answer = 0;
    
    queue<pair<int,int>> q;
    q.push({0,0});
    
    int visited[102][102];memset(visited, -1, sizeof visited);
    visited[0][0] = 1;
    int n = maps.size(), m = maps[0].size();
    
    while(!q.empty()){
        auto[x,y] = q.front(); q.pop();
        for(int i{};i<4;i++){
            int nx = x + dx[i];
            int ny = y + dy[i];
            if(nx < 0 || nx >= n || ny < 0 || ny >= m)continue;
            if(visited[nx][ny]!=-1 || !maps[nx][ny])continue;
            visited[nx][ny] = visited[x][y] + 1;
            q.push({nx,ny});
        }
    }
    return visited[n-1][m-1];
}

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
'''
'25. 09. 23.(화)

bfs를 이용하여 움직이면 제일 빠르게 최단거리로 나갈 수 있다.

1. 0은 벽이 있는 자리, 1은 없는 자리이다.
2. 탈출할 수 있으면 answer, 아니면 -1을 리턴한다.
'''
from collections import deque
def bfs(n: int, m: int, board: list) -> int:
    #초기 위치 설정
    queue = deque([[0, 0]])
    
    #방문 배열
    visited = [[-1] * m for _ in range(n)]
    MOVE = [[0, 1], [0, -1], [1, 0], [-1, 0]]
    
    #처음 위치 0
    visited[0][0] = 1
    while queue:
        x, y = queue.popleft()
        #탈출하였음
        if x == n - 1 and y == m - 1:
            return visited[x][y]
        for dx, dy in MOVE:
            nx, ny = x + dx, y + dy
            if nx < 0 or nx >= n or ny < 0 or ny >= m:continue
            if visited[nx][ny]!= -1 or board[nx][ny] == 0:continue
            visited[nx][ny] = visited[x][y] + 1
            queue.append([nx, ny])
    #마지막에 탈출하지 못 하였을 때
    return -1
def solution(maps):
    return bfs(len(maps), len(maps[0]), maps)